Patch the make_rst.py utility to handle specially operators with "<"

jynus opened the pull request:
When generating rst files from xml class reference, unknown references
to operators were generated, as something like:

:ref:`operator <<class_Vector2_operator_lt_bool>`

was rendered in html as:

operator ( Vector2 right )

-it just needed escaping.

The small addendum checks for operator names containing '<' and
substitutes it with '\<', escaping at rst level and generating
instead the right rendered html:

operator < ( Vector2 right )

This affected mostly the reference pages of the VectorX family of
classes. If in the future more types need escaping, a more
general solution will be needed.

jynus commented:
Before: https://docs.godotengine.org/en/latest/classes/class_vector2.html#operators
![before](https://user-images.githubusercontent.com/8576860/180661089-b15eeb2c-1684-4512-9632-0fbd51a41d64.png)

and after:
![after](https://user-images.githubusercontent.com/8576860/180661092-3892e699-1939-43c5-a740-f9ecc07ba6ff.png)

jynus commented:
I thought < was an escapable character, that is why I initially used the single quotes. I just realized it is a normal one for Python, so "\<" worked as expected. The script doesn't seem super-complex, so a small patch may be enough here- just for operators?

YuriSizov commented:
Maybe the display text should simply be HTML encoded, so it's turned into something like `&lt;` or a unicode number?

jynus commented:
Thanks for the comments, I will give it a second pass.

jynus commented:
So I've gone into the rabbit hole of safe rst and I reached a few conclusions:

* We cannot HTML-encode the strings- e.g. if you write &lt; on rst it will translate it to &amp;lt;
* Even if it worked, HTML is not the right level to fix this- we want to fix having syntactically-valid rst, not just the final output HTML. For example, if we wanted to export to PDF from RST, we need the source rst docs to be valid.
* There is not generally-used sanitizing function for rst text- unlike bash, sql, html and other languages. This is for 2 reasons: 1) rst is intended for human writing, so you normally get people saying "just escape the badly-behaving character with a backslash" 2) rst is context-aware, so the same set of characters can be interpreted or not as rst special syntax in a very non-deterministic way- similar to HTML, we would need to reparse the entire page for each literal, but even then; that may not work because...
* I found an escaping implementation on sphinx source code: https://github.com/sphinx-doc/sphinx/tree/5.x/sphinx/util#L27 but that doesn't even work for us, because rst/sphinx can be configured/added plugins that make it not enough for our needs (doesn't account for all special characters)
* That doesn't mean we have to approve the original patch, I have written a function based on the above that looks like this:

```python3
def sanitize_rst_literal(text):  # type: (str) -> str
    for character in '\\!"#$''()*+-.:<=>[]^_`{|}~':
        if character in text:
            text = text.replace(character, "\\" + character)
    return text
```
and using sparsingly around the code (as other part already return rst directly) should work ok for our usage, and it has been optimized for performance (actually using an array is +2% faster, and used static typing a +0.1%), but I think this version is preferred (yes I did several benchmarks and unit tests for this).

I will send the updated PR when I do a diff of all the documentation before and after the patch for validation. Unit testing would be great for this file, which I could add on a separate PR (alongside refactoring it with modern python). While keeping software compatible with older versions should be an aim, I feel that for what accounts as internal tooling -which won't be normally run by end users- could have higher requirements (but that is a discussion for another time).

YuriSizov commented:
@jynus Thanks for investigating this! If this is the case, I think the original solution is fine and sufficient. No need to make it more complex for the sake of having a more flexible tool for something that isn't a problem yet 🙂 

If you can only fix the comment style, I think this is good as is.

jynus commented:
It is ok, now that I have done the work let me send the "proper" way (with a separate function) at least :-D

jynus commented:
Let's see if CI is happy- now there is some extra escaping on method names e.g.:

```
ref:`append\_array<class_PackedVector2Array_method_append_array>`
```

But as far as I can see all diff look as expected on rst, and links/appeareance looks good on html.

jynus commented:
Never been a fan of black- the version that it likes is 4-6% slower, but it is what it is :-D.

jynus commented:
Please try the output and let me know if the fix is ok for you, too.

YuriSizov commented on doc/tools/make_rst.py:
```diff
@@ -1417,7 +1433,7 @@ def make_method_signature(
         # See `make_footer()` for descriptions.
         for qualifier in method_def.qualifiers.split():
-            out += " |" + qualifier + "|"
+            out += " |" + sanitize_rst_literal(qualifier) + "|"
```
Don't think we need to escape qualifiers, those are just text that we have predefined.

YuriSizov reviewed (approved):
Seems to work, tried to rebuild online docs locally with it, and it was all fine.

jynus commented:
I sent an update with your latest suggestion.

akien-mga commented:
> Let's see if CI is happy- now there is some extra escaping on method names e.g.:
> 
> ```
> ref:`append\_array<class_PackedVector2Array_method_append_array>`
> ```
> 
> But as far as I can see all diff look as expected on rst, and links/appeareance looks good on html.

I'm not so fond of that part. Do we ever need to escape underscores and dots this way?

This generates a lot of diff and extra escaping and if we don't actually need it, I'd prefer not to do it.
E.g.
```diff
 
 .. _class_@GlobalScope_method_tan:
 
-- :ref:`float<class_float>` **tan** **(** :ref:`float<class_float>` angle_rad **)**
+- :ref:`float<class_float>` **tan** **(** :ref:`float<class_float>` angle\_rad **)**
 
 Returns the tangent of angle ``angle_rad`` in radians.
 
@@ -4198,7 +4198,7 @@ Returns the hyperbolic tangent of ``x``.
 
 - :ref:`int<class_int>` **typeof** **(** :ref:`Variant<class_Variant>` variable **)**
 
-Returns the internal type of the given Variant object, using the :ref:`Variant.Type<enum_@GlobalScope_Variant.Type>` values.
+Returns the internal type of the given Variant object, using the :ref:`Variant\.Type<enum_@GlobalScope_Variant.Type>` values.
 
 ::
```

The `Variant.Type` reference seems to work fine before this PR: https://docs.godotengine.org/en/latest/classes/class_%40globalscope.html#class-globalscope-method-typeof
And methods or arguments with underscores have never been an issue.

akien-mga commented:
So adding to my above comment, I would suggest to change the validation method to only escape what we know is currently problematic. We can always add more in the future if we see a need to escape more characters.

jynus commented:
Could @YuriSizov weight in on the scope? My first patch suggestion was actually what @akien-mga suggests now (a very narrow fix/small patch for what we know is broken- a replace()), but I increased it upon [Yuri's request](#issuecomment-1193376594) to make escaping more globaly-applyable. :-)

I am happy to leave it either way, as long as the current broken docs are fixed and all people agree. 0:-)

jynus commented:
> Remember that we do write some references manually sometimes when writing tutorials which should refer to a specific class/method/property. So they need to stay human readable and predictable:
> 
> SomeClass.some_method -> :ref:`class_SomeClass_method_some_method`

As a note -as there seems to be a confusion-, reference identifiers were **not** touched in the latest version, only plain text identifiers (e.g. link text was escaped, never link address).

YuriSizov commented:
> Could @YuriSizov weight in on the scope? My first patch suggestion was actually what @akien-mga suggests now (a very narrow fix/small patch for what we know is broken- a replace()), but I increased it upon [Yuri's request](#issuecomment-1193376594) to make escaping more globaly-applyable. :-)

Oh, I agree with Akien. That's what I meant here: 

> If this is the case, I think the original solution is fine and sufficient. No need to make it more complex for the sake of having a more flexible tool for something that isn't a problem yet 🙂
> 
> If you can only fix the comment style, I think this is good as is.

Since there was no standardized way of escaping characters, then we can stop at escaping just the lesser than sign, like you initially did. I just didn't mind much that you've decided to implement a more robust fix for potential future issues as well. 🙂

jynus commented:
(sorry, by rebasing using GitHub web interface rather than cmd line, it closed automatically the pull request- unintentionally)

jynus commented:
Only operators affected on this latest version: [classes.diff.txt](https://github.com/godotengine/godot/files/9208127/classes.diff.txt) :-)

akien-mga commented:
I think you didn't push the new commit yet, it's still using the full blown list IINM.

jynus commented:
This is the intended latest version: https://github.com/godotengine/godot/pull/63400/commits/9a7fffd396081edb63e27aaf9a0876a3cd467a19 It does full escaping, but applies only to operators.

Should I just go to the very original patch instead? https://github.com/godotengine/godot/commit/2fb26234666506aef742b840448dcecb38df9f77

akien-mga commented:
Yes I would go with 2fb26234666506aef742b840448dcecb38df9f77 if that's the only change needed.

Adding a general purpose method to sanitize identifers but then only use it for a very specific case doesn't seem useful to me. Either we apply it to everything that could fail in the future (future proofing) but only for the actual characters that can cause a problem, or we only do a fix where it's actually needed (keep solutions local). [The latter is usually our preferred guideline.](https://docs.godotengine.org/en/stable/community/contributing/best_practices_for_engine_contributors.html)

Sorry if the successive reviews sent you through a rabbit hole of generic fixing unnecessarily.

jynus commented:
It is super ok- in a way validates initial submission as the "righ way"- so thank you! Doing it- with some small formatting fixes for black.

jynus committed:
Patch the make_rst.py utility to handle specially operators with '<'

When generating rst files from xml class reference, unknown references
to operators were generated, as something like:

:ref:`operator <<class_Vector2_operator_lt_bool>`

was rendered in html as:

operator ( Vector2 right )

-it just needed escaping.

The small addendum checks for operator names containing '<' and
substitutes it with '\<', escaping at rst level and generating
instead the right rendered html:

operator < ( Vector2 right )

This affected mostly the reference pages of the VectorX family of
classes. If in the future more types need escaping, a more
general solution will be needed.

jynus commented:
Latest minimal diff: only operators with "<":
[classes.diff.txt](https://github.com/godotengine/godot/files/9208318/classes.diff.txt)

akien-mga commented:
Thanks!
